core/capabilities/remote/executable/request/payload_comparator.go

```diff
@@ -0,0 +1,155 @@
+)
+
+type ComparatorState struct {
+	Message    *types.MessageBody
+	Metadata   commoncap.ResponseMetadata
+	Hash       [32]byte
+	ResponseID string // The hex-encoded hash string
+}
+
+// payloadComparator handles logging and comparing divergent payloads.
+type payloadComparator struct {
+	lggr       logger.Logger
+	firstState atomic.Pointer[ComparatorState]
+}
+
+// NewPayloadComparator creates and initializes a new comparator.
+func NewPayloadComparator(lggr logger.Logger) *payloadComparator {
+	return &payloadComparator{
+		lggr: lggr,
+	}
+}
+
+// Compare accepts an incoming message, stores the first unique instance, and logs
+// differences against subsequent unique messages.
+func (pc *payloadComparator) Compare(incomingMsg *types.MessageBody) {
+	incomingHash, md, err := GetMessageHashAndMetadata(pc.lggr, incomingMsg)
+	if err != nil {
+		pc.lggr.Errorw("failed to get message data and hash", "error", err)
+		return
+	}
+
+	if pc.firstState.Load() == nil {
+		newState := &ComparatorState{
+			Message:    incomingMsg,
+			Metadata:   md,
+			Hash:       incomingHash,
+			ResponseID: hex.EncodeToString(incomingHash[:]),
+		}
+
+		if !pc.firstState.CompareAndSwap(nil, newState) {
+			pc.lggr.Error("failed to load first state in comparator")
+			return
+		} else {
+			pc.lggr.Debugw("Comparator stored first unique response payload.",
+				"responseID", newState.ResponseID,
+				"metadata", newState.Metadata,
+			)
+		}
+	}
+
+	existingState := pc.firstState.Load()
+	if bytes.Equal(existingState.Hash[:], incomingHash[:]) {
+		return
+	}
+
+	pc.lggr.Warnw("Divergent response hash detected. Running detailed payload comparison.",
+		"firstResponseID", existingState.ResponseID,
+		"firstMetadata", existingState.Metadata,
+		"incomingResponseID", hex.EncodeToString(incomingHash[:]),
+		"incomingMetadata", md,
+	)
+
+	incomingResp, err := pb.UnmarshalCapabilityResponse(incomingMsg.Payload)
+	if err != nil {
+		pc.lggr.Warnw("failed to unmarshal incoming message", err)
+		return
+	}
+
+	expectedResp, err := pb.UnmarshalCapabilityResponse(existingState.Message.Payload)
+	if err != nil {
+		pc.lggr.Warnw("failed to unmarshal first message", err)
+		return
+	}
+
+	logAnyPayloadByteDiffs(pc.lggr, incomingResp, expectedResp, incomingHash, existingState.Hash)
+}
+
+func logAnyPayloadByteDiffs(lggr logger.Logger, respA, respB commoncap.CapabilityResponse, hashA, hashB [32]byte) {
+	if (respA.Payload == nil) != (respB.Payload == nil) {
+		lggr.Errorw("Payload divergence: One response has a Payload and the other doesn't",
+			"hashA_has_payload", respA.Payload != nil,
+			"hashB_has_payload", respB.Payload != nil,
+		)
+		return
+	}
+	if respA.Payload == nil && respB.Payload == nil {
+		return
+	}
+
+	if proto.Equal(respA.Payload, respB.Payload) {
+		return
+	}
+
+	valA := respA.Payload.GetValue()
+	valB := respB.Payload.GetValue()
+
+	lenA := len(valA)
+	lenB := len(valB)
+	minLen := lenA
+	if lenB < minLen {
+		minLen = lenB
+	}
```

Comment:
Min is built in now:
```suggestion
	minLen := min(lenA, lenB)
```